File: 1021-remove-outermost-parentheses/1021-remove-outermost-parentheses.java
class Solution {
    public String removeOuterParentheses(String s) {
        int len = s.length();
        int pL = 0;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i <= len; i++) {
            String sub = s.substring(pL, i);
            if (isValidParentheses(sub)) {
                sb.append(sub.substring(1, sub.length() - 1));
                pL = i;
            }
        }
        return sb.toString();
    }
    
    private boolean isValidParentheses(String s) {
        if (s.isBlank()) return false;
        char[] chars = s.toCharArray();
        Stack<Character> stack = new Stack<>();
        for (char c : chars) {
            if (!stack.isEmpty() && c == ')') {
                if (stack.peek() == '(') stack.pop();
            } else {
                stack.add(c);
            }
        }    
        return stack.size() == 0;
    }
}

File: 1021-remove-outermost-parentheses/1021-remove-outermost-parentheses.kt
import java.util.*


class Solution {
    fun removeOuterParentheses(s: String): String {

        val stack = Stack<Char>()
        val list: MutableList<Int> = ArrayList()
        val sb = StringBuilder()

        for (i in s.indices) {
            val currentChar: Char = s[i]
            if (stack.isEmpty()) {
                list.add(i)
                stack.push(currentChar)
            } else if (stack.peek() == '(' && currentChar == ')' || stack.peek() == ')' && currentChar == '(') {
                stack.pop()
                if (stack.isEmpty()) {
                    list.add(i)
                }
            } else {
                stack.push(currentChar)
            }
        }
        var i = 0
        while (i < list.size) {
            val startIdx = list[i]
            val endIdx = list[i + 1]
            sb.append(s.substring(startIdx + 1, endIdx))
            i += 2
        }


        return sb.toString()

    }
}